File: app/src/androidTest/java/com/jbrunton/mymovies/fixtures/BaseTest.java
package com.jbrunton.mymovies.fixtures;

import android.app.Activity;
import android.support.test.rule.ActivityTestRule;

import com.google.android.libraries.cloudtesting.screenshots.ScreenShotter;
import com.squareup.spoon.SpoonRule;

import org.junit.Rule;

public abstract class BaseTest<T extends Activity> {
    @Rule public final ActivityTestRule<T> activityRule = createActivityTestRule();
    @Rule public final SpoonRule spoonRule = new SpoonRule();

    public void takeScreenshot() {
        takeScreenshot("_");
    }

    public void takeScreenshot(String tag) {
        spoonRule.screenshot(activityRule.getActivity(), tag);
        ScreenShotter.takeScreenshot(tag, activityRule.getActivity());
    }

    protected abstract ActivityTestRule<T> createActivityTestRule();
}

File: app/src/androidTest/java/com/jbrunton/mymovies/fixtures/BaseTest.kt
package com.jbrunton.mymovies.fixtures

import android.app.Activity
import androidx.test.rule.ActivityTestRule

import com.google.android.libraries.cloudtesting.screenshots.ScreenShotter
import com.squareup.spoon.SpoonRule

import org.junit.Rule
import org.koin.test.KoinTest

abstract class BaseTest<T : Activity> : KoinTest {
    @get:Rule
    val activityRule = createActivityTestRule()
    @get:Rule
    val spoonRule = SpoonRule()

    @JvmOverloads
    fun takeScreenshot(tag: String = "_") {
        // TODO: figure out why getting permissions errors since upgrading to using API 28
        // spoonRule.screenshot(activityRule.activity, tag)
        ScreenShotter.takeScreenshot(tag, activityRule.activity)
    }

    protected abstract fun createActivityTestRule(): ActivityTestRule<T>
}
